libexec/rtld-elf/rtld_c18n.c

```diff
@@ -0,0 +1,770 @@
+ */
+struct tramp_pg {
+	SLIST_ENTRY(tramp_pg) link;
+	_Atomic(size_t) size;
+	size_t capacity;
+	_Alignas(_Alignof(void *)) char trampolines[];
+};
+
+/*
+* 64K is the largest size such that any capability-aligned proper
+* sub-range can be exactly bounded by a Morello capability.
+*/
+#define	C18N_TRAMPOLINE_PAGE_SIZE	64 * 1024
+
+static struct tramp_pg *
+tramp_pg_new(struct tramp_pg *next)
+{
+	size_t capacity = C18N_TRAMPOLINE_PAGE_SIZE;
+	struct tramp_pg *pg = mmap(NULL,
+	    capacity,
+	    PROT_READ | PROT_WRITE | PROT_EXEC,
+	    MAP_ANON | MAP_PRIVATE,
+	    -1, 0);
+	if (pg == MAP_FAILED)
+		rtld_fatal("mmap failed");
+	SLIST_NEXT(pg, link) = next;
+	atomic_store_explicit(&pg->size, 0, memory_order_relaxed);
+	pg->capacity = capacity - offsetof(typeof(*pg), trampolines);
+	return (pg);
+}
+
+static void *
+tramp_pg_push(struct tramp_pg *pg, size_t len)
+{
+	void *tramp;
+	size_t n_size;
+	size_t size = atomic_load_explicit(&pg->size, memory_order_relaxed);
+	do {
+		n_size = roundup2(size, _Alignof(void *));
+		tramp = pg->trampolines + n_size;
+		n_size += len;
+		if (n_size > pg->capacity)
+			return (NULL);
+	} while (!atomic_compare_exchange_weak_explicit(&pg->size, &size, n_size,
+	    /*
+	     * Relaxed ordering is sufficient because there are no
+	     * side-effects.
+	     */
+	    memory_order_relaxed, memory_order_relaxed));
+
+	tramp = cheri_setbounds(tramp, len);
+	assert(cheri_gettag(tramp));
+
+	return (tramp);
+}
+
+typedef ssize_t slot_idx_t;
+
+static struct {
+	struct tramp_data *data;
+	struct tramp_map_kv {
+		_Atomic(ptraddr_t) key;
+		_Atomic(slot_idx_t) index;
+	} *map;
+	_Atomic(slot_idx_t) size;
+	_Atomic(slot_idx_t) writers;
+	int exp;
+} tramp_table;
+
+static struct {
+	_Atomic(struct tramp_pg *) head;
+	atomic_flag lock;
+} tramp_pgs = {
+	.lock = ATOMIC_FLAG_INIT
+};
+
+static slot_idx_t
+tramp_table_max_load(int exp)
+{
+	/* MAX_LOAD is 37.5% of capacity. */
+	return (3 << (exp - 3));
+}
+
+static struct tramp_map_kv *
+tramp_map_new(int exp)
+{
+	struct tramp_map_kv *map;
+
+	map = xmalloc(sizeof(*map) << exp);
+	for (size_t i = 0; i < (1 << exp); ++i)
+		map[i] = (struct tramp_map_kv) {
+			.key = 0,
+			.index = -1
+		};
+	return (map);
+}
+
+static struct tramp_data *
+tramp_data_expand(int exp)
+{
+	struct tramp_data *data;
+	/*
+	 * The table only needs to be as large as the MAX_LOAD.
+	 */
+	data = realloc(tramp_table.data, sizeof(*data) * tramp_table_max_load(exp));
+	if (data == NULL)
+		rtld_fatal("realloc failed");
+	return (data);
+}
+
+/* Public domain. Taken from https://github.com/skeeto/hash-prospector */
+static uint32_t
+pointer_hash(uint64_t key)
+{
+	uint32_t x = key ^ (key >> 32);
+	x ^= x >> 16;
+	x *= 0x21f0aaadU;
+	x ^= x >> 15;
+	x *= 0xd35a2d97U;
+	x ^= x >> 15;
+	return (x);
+}
+
+static slot_idx_t
+nextSlot(slot_idx_t slot, uint32_t hash, int exp)
+{
+	uint32_t mask = (1 << exp) - 1;
+	uint32_t step = (hash >> (32 - exp)) | 1;
+	return ((slot + step) & mask);
```

Comment:
```suggestion

	return ((slot + step) & mask);
```

Reply:
Though this might look better as:
```C
	uint32_t mask, step;

	mask = (1 << exp) - 1;
 	step = (hash >> (32 - exp)) | 1;
 	return ((slot + step) & mask);
```